src/rust/aggregator/src/pipelines/order_history_pipelines.rs

```diff
@@ -0,0 +1,594 @@
+                transaction = calculate_spreads_and_liquidities(
+                    &address_to_group,
+                    &market_id_to_group,
+                    &mut self.last_prices,
+                    state,
+                    timestamp,
+                    transaction,
+                )
+                .await?;
+            }
+        }
+        {
+            if let Some(last) = timestamps.last() {
+                sqlx::query_file!(
+                    "sqlx_queries/order_history_pipelines/update_last_indexed_timestamp.sql",
+                    last,
+                )
+                .execute(&mut transaction as &mut PgConnection)
+                .await
+                .map_err(to_pipeline_error)?;
+            }
+        }
+        transaction.commit().await.map_err(to_pipeline_error)?;
+        Ok(())
+    }
+}
+
+async fn calculate_spreads_and_liquidities(
+    address_to_group: &HashMap<String, i32>,
+    market_id_to_group: &HashMap<BigDecimal, i32>,
+    last_prices: &mut HashMap<BigDecimal, Option<BigDecimal>>,
+    state: &SpreadsState,
+    timestamp: &DateTime<Utc>,
+    transaction: Transaction<'static, Postgres>,
+) -> Result<Transaction<'static, Postgres>, PipelineError> {
+    let tx = Arc::new(RwLock::new(transaction));
+    let (res1, res2) = tokio::join! {
+        liquidities(&address_to_group, &market_id_to_group, last_prices, state, timestamp, tx.clone()),
+        spreads(state, timestamp, tx.clone())
+    };
+    res1?;
+    res2?;
+    Ok(Arc::into_inner(tx).unwrap().into_inner())
+}
+
+async fn spreads<'a>(
+    state: &SpreadsState,
+    timestamp: &DateTime<Utc>,
+    transaction: Arc<RwLock<Transaction<'a, Postgres>>>,
+) -> Result<(), PipelineError> {
+    let (min_asks, max_bids) = get_spreads_from_state(state);
+    let mut markets = min_asks.keys().chain(max_bids.keys()).collect::<Vec<_>>();
+    markets.sort();
+    markets.dedup();
+    for market in markets {
+        let mut tx = transaction.write().await;
+        sqlx::query_file!(
+            "sqlx_queries/order_history_pipelines/insert_spread.sql",
+            market,
+            timestamp,
+            min_asks.get(market),
+            max_bids.get(market),
+        )
+        .execute(&mut tx as &mut PgConnection)
+        .await
+        .map_err(to_pipeline_error)?;
+    }
+    Ok(())
+}
+
+async fn liquidities<'a>(
+    address_to_group: &HashMap<String, i32>,
+    market_id_to_group: &HashMap<BigDecimal, i32>,
+    last_prices: &mut HashMap<BigDecimal, Option<BigDecimal>>,
+    state: &SpreadsState,
+    timestamp: &DateTime<Utc>,
+    transaction: Arc<RwLock<Transaction<'a, Postgres>>>,
+) -> Result<(), PipelineError> {
+    let liquidities = get_liquidities_from_state(
+        address_to_group,
+        market_id_to_group,
+        &transaction,
+        state,
+        last_prices,
+    )
+    .await?;
+    for (key, value) in liquidities {
+        let mut tx = transaction.write().await;
```

Comment:
So I think that we should use the repeatable txn helper at the beginning of the main loop and just pass references to it

That's what we do in the other pipelines and the approach here is nonstandard